components/calendar/calendar.js

```diff
@@ -0,0 +1,488 @@
+	let nextMonthDay = 1;
+	let firstDateOfWeek = daysInWeek - numDaysFromPrevMonthToShow + 1;
+	const numWeeks = Math.ceil((numDaysFromPrevMonthToShow + numDays) / daysInWeek);
+	const nextMonth =  getNextMonth(shownMonth);
+	const nextMonthYear = (shownMonth === 11) ? (shownYear + 1) : shownYear;
+	for (let i = 1; i < numWeeks; i++) {
+		const week = [];
+		for (let j = firstDateOfWeek; j < firstDateOfWeek + daysInWeek; j++) {
+			let day;
+			if (j > numDays) {
+				day = new Date(nextMonthYear, nextMonth, nextMonthDay);
+				nextMonthDay++;
+			} else {
+				day = new Date(shownYear, shownMonth, j);
+			}
+			week.push(day);
+		}
+		firstDateOfWeek = firstDateOfWeek + daysInWeek;
+		dates.push(week);
+	}
+
+	return dates;
+}
+
+export function getNextMonth(month) {
+	return (month === 11) ? 0 : (month + 1);
+}
+
+export function getNumberOfDaysFromPrevMonthToShow(month, year) {
+	const firstDayOfMonth = new Date(year, month, 1).getDay();
+	let numDaysFromLastMonthToShowThisMonth = 0;
+	if (firstDayOfMonth !== firstDayOfWeek) {
+		numDaysFromLastMonthToShowThisMonth = firstDayOfMonth - firstDayOfWeek;
+		if (numDaysFromLastMonthToShowThisMonth < 0) {
+			numDaysFromLastMonthToShowThisMonth += daysInWeek;
+		}
+	}
+	return numDaysFromLastMonthToShowThisMonth;
+}
+
+export function getNumberOfDaysInMonth(month, year) {
+	return new Date(year, month + 1, 0).getDate();
+}
+
+export function getNumberOfDaysToSameWeekPrevMonth(month, year) {
+	const prevMonth = getPrevMonth(month);
+	const prevMonthYear = prevMonth === 11 ? year - 1 : year;
+	const numDaysPrevMonth = getNumberOfDaysInMonth(prevMonth, prevMonthYear);
+	const numDaysFromPrevPrevMonthToShowPrevMonth = getNumberOfDaysFromPrevMonthToShow(prevMonth, prevMonthYear);
+	const numWeeksPrevMonth = Math.ceil((numDaysFromPrevPrevMonthToShowPrevMonth + numDaysPrevMonth) / 7);
+
+	const firstDayOfMonth = new Date(year, month, 1).getDay();
+	const hasDaysFromPrevMonth = (firstDayOfMonth !== firstDayOfWeek);
+	return daysInWeek * (numWeeksPrevMonth - ((hasDaysFromPrevMonth) ? 1 : 0));
+}
+
+export function getPrevMonth(month) {
+	return (month === 0) ? 11 : (month - 1);
+}
+
+export function getToday() {
+	const val = new Date().toISOString();
+	const re = /([0-9]{4})-([0-9]{2})-([0-9]{2})T([0-9]{1,2}):([0-9]{1,2}):([0-9]{1,2})/;
+	const match = val.match(re);
+	const dateTime = {
+		year: parseInt(match[1]),
+		month: parseInt(match[2]),
+		date: parseInt(match[3]),
+		hours: parseInt(match[4]),
+		minutes: parseInt(match[5]),
+		seconds: parseInt(match[6])
+	};
+
+	const valInLocalDateTime = convertUTCToLocalDateTime(dateTime);
+	return new Date(valInLocalDateTime.year, valInLocalDateTime.month - 1, valInLocalDateTime.date);
+}
+
+export function parseDate(val) {
+	if (!val) return null;
+	const re = /([0-9]{4})-([0-9]{2})-([0-9]{2})/;
+	const match = val.match(re);
+	if (!match || match.length !== 4) {
+		throw new Error('Invalid selected-value date input: Expected format is YYYY-MM-DD');
```

Comment:
Since this can throw, does the code that calls this need to handle the exceptions at all, or is it mostly dev-user error cases where you want them to see the exceptions? Basically I want to confirm that there's no way for end users to cause errors in the console, only developers.

Reply:
My thought was that it should only be the dev-user error cases. Potentially we could deal with them, though I think it would be important for devs to know that they did something wrong.

Reply:
Cool I'm good with throwing then if it's only dev-user error :)